### PR TITLE
[fix][ML] Fix offload read handle NPE.

### PR DESCRIPTION
**Pull request by horizonzy**
### Motivation
Now, when we get offload ReadHandle(`BlobStoreBackedReadHandleImpl`) and then use it to read data (Non-durable cursor read). Maybe the offload ReadHandle we get already is invalidated.

There is a race condition, maybe after we get the ReadHandle, before use it to read. 
The method `ManagedLedgerImpl#internalTrimLedgers` invalidate this ReadHandle (Cause in this method, it just calculate by durable-cursor, but there maybe a non-durable cursor read data) so we use the invalidated offload read handle to read data, invalidated will close the read handle. 

https://github.com/apache/pulsar/blob/96930fda8a7776aaf6e61235a1d77ffe6e564052/managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java#L2542

In the close method, it will set `index.segmentMetadata=null`, so when getLastAddConfirmed, case NPE.
https://github.com/apache/pulsar/blob/bc68f02a4060ef0042a50c2b1d834c6ab84cfd13/tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImpl.java#L100

**The NPE stack info**
```
rg.apache.bookkeeper.mledger.ManagedLedgerException: java.lang.NullPointerException
Caused by: java.lang.NullPointerException
	at org.apache.bookkeeper.mledger.offload.jcloud.impl.BlobStoreBackedReadHandleImpl.getLastAddConfirmed(BlobStoreBackedReadHandleImpl.java:200) ~[?:?]
	at org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.internalReadFromLedger(ManagedLedgerImpl.java:1870) ~[io.streamnative-managed-ledger-2.9.2.23.jar:2.9.2.23]
	at org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.lambda$asyncReadEntries$17(ManagedLedgerImpl.java:1719) ~[io.streamnative-managed-ledger-2.9.2.23.jar:2.9.2.23]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:714) ~[?:?]
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2073) ~[?:?]
	at org.apache.bookkeeper.mledger.impl.ManagedLedgerImpl.lambda$null$20(ManagedLedgerImpl.java:1790) ~[io.streamnative-managed-ledger-2.9.2.23.jar:2.9.2.23]
	at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) [?:?]
	at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) [?:?]
	at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:478) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) [io.netty-netty-common-4.1.77.Final.jar:4.1.77.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
03:30:51.930 [broker-topic-workers-OrderedExecutor-3-0] ERROR org.apache.pulsar.broker.service.persistent.PersistentDispatcherSingleActiveConsumer - [persistent://public/default/test / reader-51d0ba5f20-Consumer{subscription=PersistentSubscription{topic=persistent://public/default/test, name=reader-51d0ba5f20}, consumerId=0, consumerName=reader-0, address=/10.244.1.143:51782}] Error reading entries at 0:101 : java.lang.NullPointerException - Retrying to read in 15.0 seconds

```

### Modifications
1.Fix the NPE.
2.Check the invalidate ledgerId is less than the slowest non-durable, reduce the npe probability.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -2326,14 +2328,27 @@ public void checkInactiveSubscriptions() {
...
+            }
             if (!subscription.getConsumers().isEmpty()
                 && subscription.getCursor().getNumberOfEntries() < backloggedCursorThresholdEntries) {
                 subscription.getCursor().setActive();
             } else {
                 subscription.getCursor().setInactive();
             }
         });
+        if (slowestNonDurableReadPosition.get() != null) {
+            ManagedLedger managedLedger = getManagedLedger();
```
Comment (hangc0276) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2349]: You can directly use `ledger.updateTheSlowestNonDurableReadPosition`

**Review by hangc0276 — APPROVED** [review]
Nice catch! LGTM.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java [hunk 2]
```diff
@@ -791,7 +791,8 @@ public synchronized void readEntriesFailed(ManagedLedgerException exception, Obj
                 // Notify the consumer only if all the messages were already acknowledged
                 consumerList.forEach(Consumer::reachedEndOfTopic);
             }
-        } else if (exception.getCause() instanceof TransactionBufferException.TransactionNotSealedException) {
+        } else if (exception.getCause() instanceof TransactionBufferException.TransactionNotSealedException
+                || exception.getCause() instanceof ManagedLedgerException.OffloadReadHandleClosedException) {
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:795]: Looks like we don't want to handle the race condition of trimming ledgers and reading data, instead, to re-trigger the read operation?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2349]: +1

**Comment by codelipenghui** [comment]
And can we add a test to make sure the reader handle will not invalidate if we the slowest read position haven't go through it yet.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 2 above)
Comment (horizonzy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:795]: Yes, it's difficult to handle the race condition, the offload read handle execute read operation in offload-executor, before the read real execute, the read handle maybe invalidated.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (horizonzy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:2349]: Agrre, and there will reset `ManagedLedgerImpl.slowestNonDurableReadPosition` when there is no non-duranble cursor.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java (hunk 2 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:795]: +1 for me.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java [hunk 3]
```diff
@@ -737,6 +738,69 @@ public void testBacklogStatsWhenDroppingData() throws Exception {
...
+
+        CompletableFuture<Void> promise = new CompletableFuture<>();
+        ledger.internalTrimConsumedLedgers(promise);
+        promise.join();
+
+        Assert.assertTrue(ledger.ledgerCache.containsKey(3));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(4));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(5));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(6));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(7));
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:785]: We should avoid to use the hardcode ledger ID.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java [hunk 4]
```diff
@@ -737,6 +738,69 @@ public void testBacklogStatsWhenDroppingData() throws Exception {
...
+
+        promise = new CompletableFuture<>();
+
+        nonDurableCursor.markDelete(positions.get(3));
+        ledger.updateTheSlowestNonDurableReadPosition(nonDurableCursor.getReadPosition());
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:790]: Do we need this line? Looks the read position will not change.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java [hunk 5]
```diff
@@ -737,6 +738,69 @@ public void testBacklogStatsWhenDroppingData() throws Exception {
         ledger.close();
     }
 
+    @Test
+    public void testInvalidateReadHandleWithSlowNonDurableCursor() throws Exception {
+        ManagedLedgerImpl ledger = (ManagedLedgerImpl) factory.open("testInvalidateReadHandleWithSlowNonDurableCursor",
+                new ManagedLedgerConfig().setMaxEntriesPerLedger(1).setRetentionTime(-1, TimeUnit.SECONDS)
+                        .setRetentionSizeInMB(-1));
+        ManagedCursor c1 = ledger.openCursor("c1");
+        ManagedCursor nonDurableCursor = ledger.newNonDurableCursor(PositionImpl.EARLIEST);
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:747]: Looks like we don't need a real non-durable cursor?
Just update the `SlowestNonDurableReadPosition`

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java [hunk 6]
```diff
@@ -737,6 +738,69 @@ public void testBacklogStatsWhenDroppingData() throws Exception {
...
+
+        ledger.internalTrimConsumedLedgers(promise);
+        promise.join();
+
+        Assert.assertFalse(ledger.ledgerCache.containsKey(3));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(4));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(5));
+        Assert.assertFalse(ledger.ledgerCache.containsKey(6));
+        Assert.assertTrue(ledger.ledgerCache.containsKey(7));
+
```
Comment (codelipenghui) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:800]: And we should continue to update the `SlowestNonDurableReadPosition` to make sure the read handle cache can be invalidated

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java (hunk 4 above)
Comment (horizonzy) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:790]: To mock the real situation, so here define a non-durable cursor, and mark delete non-durable cursor, then update the slowestNonDurableReadPosition by this non-durable cursor.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java (hunk 5 above)
Comment (horizonzy) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:747]: To mock the real situation

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java (hunk 6 above)
Comment (horizonzy) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:800]: Here already invalidate ledger 3,4,5,6.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java (hunk 3 above)
Comment (horizonzy) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:785]: nice suggestion.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java (hunk 6 above)
Comment (Technoboy-) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:800]: > And we should continue to update the SlowestNonDurableReadPosition to make sure the read handle cache can be invalidated
I think @codelipenghui means to do repeatedly like line 789~790.

> Here already invalidate ledger 3,4,5,6.
Why invalidate `4,5,6` ?

Comment (horizonzy) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:800]: positions[0] is `3:0`,  positions[1] is `4:0`,  positions[2] is `5:0`, positions[3] is `6:0`, positions[4] `7:0`

At line 789,790. we make non-durable cursor mark-delete to 6:0, the next read postion is `7:0`, the managed-ledger non-durable slowest read position is `7:0`. So the ledger `3,4,5,6` will be invalidate

Comment (Technoboy-) [reply managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/NonDurableCursorTest.java:800]: Ah, thanks.

**Review by Technoboy- — APPROVED** [review]

File: tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java [hunk 7]
```diff
@@ -135,13 +144,20 @@ public CompletableFuture<Void> closeAsync() {
...
+                log.warn("Reading a closed read handler. Ledger ID: {}, Read range: {}-{}",
+                        ledgerId, firstEntry, lastEntry);
+                promise.completeExceptionally(new ManagedLedgerException.OffloadReadHandleClosedException());
+                return;
+            }
+
+            if (firstEntry > lastEntry
+                    || firstEntry < 0
+                    || lastEntry > getLastAddConfirmed()) {
+                promise.completeExceptionally(new IllegalArgumentException());
```
Comment (Technoboy-) [inline tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java:158]: Should we use the same exception `BKIncorrectParameterException` with v1 or both use `IllegalArgumentException` ?

Comment (horizonzy) [reply tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BlobStoreBackedReadHandleImplV2.java:158]: nice catch

**Comment by horizonzy** [comment]
Please review this pr again, use the new way to get the slowest non-durable read position.

**Review by codelipenghui — APPROVED** [review]

**Comment by michaeljmarshall** [comment]
Heads up, cherry picking this commit to older branches broke them because of a missing import. I fixed that here: 74390d46f53125154b4116e3941a6966082c0876

**Comment by Jason918** [comment]
@horizonzy Can you help open a new PR to branch-2.10? There are a lot conflict when cherry-pick directly.

**Comment by horizonzy** [comment]
alright.

**Comment by horizonzy** [comment]
> @horizonzy Can you help open a new PR to branch-2.10? There are a lot conflict when cherry-pick directly.

#17478

**Comment by Jason918** [comment]
move `release/2.10.2` to https://github.com/apache/pulsar/pull/17478

**Pull request merged**